src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderSizeRule.cs

```diff
@@ -0,0 +1,46 @@
+﻿using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Consensus.Rules;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.Rules.ProvenHeaderRules
+{
+    /// <summary>
+    /// Rule to check if the serialized sizes of the proven header components, such as merkle proof (max 512 bytes),
+    /// signature (max 80 bytes) and coinstake (max 1,000,000 bytes), do not exceed maximum possible size allocation.
+    /// </summary>
+    /// <seealso cref="ProvenHeaderValidationConsensusRuleBase" />
+    public class ProvenHeaderSizeRule : ProvenHeaderValidationConsensusRuleBase
+    {
+        public override void Run(RuleContext context)
+        {
+            Guard.NotNull(context.ValidationContext.ChainedHeaderToValidate, nameof(context.ValidationContext.ChainedHeaderToValidate));
+
+            int height = context.ValidationContext.ChainedHeaderToValidate.Height;
+
+            if (context.SkipValidation || !this.IsProvenHeaderActivated(height))
+                return;
+
+            var header = (ProvenBlockHeader)context.ValidationContext.ChainedHeaderToValidate.Header;
+
+            if ((header.MerkleProofSize == null) || (header.MerkleProofSize > PosConsensusOptions.MaxMerkleProofSerializedSize))
```

Comment:
Do we use this any other place then in this class?
`PosConsensusOptions.MaxMerkleProofSerializedSize`

Reply:
It will be used in other PH rules in @fenix2222 's follow up prs